### ISSUE TITLE
Visual Studio: Path to debug adapter executable not specified.

### ISSUE THREAD

**Issue by jschnorr** [microsoft/PTVS#6371]
Hi,
I am struggeling with Remotely debug Python code on Linux. I am following the instructions on this link:
https://docs.microsoft.com/de-de/visualstudio/python/debugging-python-code-on-remote-linux-machines?view=vs-2019

I have the following setup:
- Windows 10 Pro installation with Visual Studio Community 2019, Version 16.8.4
- Raspberry Pi 4 8GB - up to date installation running phython3.7 with debugpy version 1.2.1
- Samba connection to the Raspberry is working

I followed the instructions on the above web-site. However when I attach to the process and start the debug process I get the following error "Path to debug adapter executable not specified." No additional hints.

I have tried two different Windows 10 systems. One Windows Pro 10 system was installed from scratch. Also I tried two newely installed Raspberry Pi4 installations with the default Raspberry Image (Debian, 2021-01-11). No change.

I am new to Visual Studio Community application. I am sure I am missing something. But I have no idea where to start my process of narrowing down the issue.

Help would be really appreciated.

Best
 Juergen

**Comment by AdamYoblick** [comment]
@int19h I think this might be a debugger issue and not PTVS specific. What do you think?

**Comment by int19h** [comment]
It sounds like it's a problem with VS not properly locating the Python interpreter to spawn debugpy; if so, that would be a VS issue. I can take a look.

**Comment by AdamYoblick** [comment]
ok thanks, I'll assign you to it then. Please let me know if you need anything from me on this.

**Comment by jschnorr** [comment]
hanks for taking a look. I am also thinking that some settings are not right at Visual Studio. Here some more datapoints of what I tried over the last days.

After rebooting all systems (especially the Raspberry) I took a look at what is running as python. So I have **not** started the actual python3 interpreter. Below the running processes.

pi@Raspb-Pi4-8G:~ $ ps -ef | grep python
pi         657     1  0 09:30 ?        00:00:00 /usr/bin/python3 /usr/share/system-config-printer/applet.py
pi        1013  1007  0 09:45 pts/1    00:00:00 grep --color=auto python

However if I attach to the process in Visual Studio i am getting attaxhed process (see screen shot). And it is of course not working.

<img width="611" alt="2021-01-28" src="https://user-images.githubusercontent.com/77958321/106120754-10647480-6157-11eb-933c-9fb6e35fdee5.png">

If I am starting on the Raspberry (python3 code.py) the processes are looking like this:

pi         657     1  0 09:30 ?        00:00:01 /usr/bin/python3 /usr/share/system-config-printer/applet.py
pi        1404  1055  6 10:49 pts/0    00:00:00 python3 RemoteTest.py
pi        1410     1  0 10:49 ?        00:00:00 /usr/bin/python3 /home/pi/.local/lib/python3.7/site-packages/debugpy/adapter --for-server 34341 --host 127.0.0.1 --port 5678 --server-access-token dbc5f14e47792d1e8d8eb77f07a78b531e7b72569cad2045fb0db93b7ec90371
pi        1426  1007  0 10:50 pts/1    00:00:00 grep --color=auto python

However there is no change in Visual Studio. This confuses me a bit.

Another datapoint. I have tried Visual Studio Code SSH approach and it works. I know you can't compare these approaches. But I wanted to share this information as well.

**Comment by int19h** [comment]
The process list in remote attach case is basically fake - it's something that PTVS needs to do because VS UI requires it, but it doesn't actually enumerate the processes running on the remote machine; it just shows a dummy process corresponding to the address/port being connected to, so that "Attach" can be clicked.

**Comment by int19h** [comment]
So the most immediate bug here is that the error message is just plain wrong. The real problem is that it couldn't connect to the socket at the specified address; but, instead of properly communicating the error, `DebugAdapterRemoteProcess.Attach` just returns null, which causes VS to show the message above.

**Comment by int19h** [comment]
@jschnorr Looking closer at your process list, it sounds like you're telling debugpy to listen on 127.0.0.1. This would prevent connections from other machines.

Most likely, you're only specifying the port and not the host in `--listen` or `debugpy.listen()`, whichever one you're using - it defaults to 127.0.0.1 for host, for security reasons. To make it listen on all interfaces, do:
```py
debugpy.listen(('0.0.0.0', 5678))
```
or via CLI
```sh
debugpy --listen 0.0.0.0:5678
```

**Comment by int19h** [comment]
@AdamYoblick Looking at this:

https://docs.microsoft.com/en-us/visualstudio/python/debugging-python-code-on-remote-linux-machines?view=vs-2019

It incorrectly suggests `debugpy.listen(5678)`, which won't work for reasons described above. The docs need to be updated to reflect this.

That aside, should we treat this issue as specifically pertaining to the misleading error message in this scenario? Since, aside from the doc bug, that's the only real problem here.

**Comment by AdamYoblick** [comment]
Yeah it sounds like an error handling bug plus a documentation issue. I can try to fix the error handling, but I'm not sure how to update the official docs.

**Comment by int19h** [comment]
While we're at it - it might be better to tackle this before the "Attach" button is even hit, by making the list of processes not quite fake.

Back when PTVS used ptvsd 3, it wasn't. More specifically, the debug server was able to process an information request and respond with metadata (like process name and PID), which was then displayed in the list. So, if you entered the URL wrong, or if nothing was listening there, you'd get an error message at that point, and "Attach" wouldn't even be enabled because there'd be nothing in the process list. 

When PTVS moved to ptvsd 5 / debugpy, this was dropped because there's no way to do a similar thing with them. Also, if I remember correctly, even simple port probing wouldn't work, because the debug adapter would simply stop running after the first connection went away.

But, I don't think that's true anymore. It should be okay to connect and then immediately disconnect now, I think. And debugpy already sends a telemetry message to the client, identifying its version:
```
D+00000.029: Client[1] <-- {
                 "seq": 2,
                 "type": "event",
                 "event": "output",
                 "body": {
                     "category": "telemetry",
                     "output": "debugpy",
                     "data": {
                         "packageVersion": "1.2.1"
                     }
                 }
             }
```
So this could be used to re-implement probing similar to how it worked in ptvsd 3 - it just needs some logic to open a DAP connection and parse incoming messages until it receives this one (it does not require any requests).

Furthermore, it would be fairly easy for debugpy to also send a custom event include more information - in particular, if it's running in attach mode managing some processes, it could include information about those processes. Then we could restore the original experience in full. I think this might be less confusing than the current behavior of always showing the dummy process, and then complaining on attach if there's no actual socket there.

**Comment by jschnorr** [comment]
This was really helpful. I was able to start the debugger but I ran into another issue. As advised I included the IP of the remote system and the debugger started. However my breakpoints are ignored!?
Maybe I am missing something. My workflow is
1) starting the actual Python script on the Raspberry. The script stops at the line debugpy.wait_for_client.
2) I start the debug process on Windows with "Attach to Process".

What happens then is: The Python scripts continues but is ignoring my breakpoints. As you can see in the screendump. If I click on one of the breakpoints the message comes up "_The breakpoint will currently not been hit. Breakpoint in file which does not exist."_

Well, the source file really exist. I am actually working directly via Samba on that file. Any suggestions?

<img width="633" alt="Screenshot" src="https://user-images.githubusercontent.com/77958321/106766438-27650400-663a-11eb-834e-7763f762f742.png">

**Comment by int19h** [comment]
This usually means that your path mappings aren't set up right - the debugger needs to know which directories on the remote machine correspond to which on local, since filenames alone are not sufficient to map.

I'm not sure how this is properly done in PTVS, though - in VSCode, it's in launch.json.

**Comment by jschnorr** [comment]
Well, I tried last night to figure out what the issue might be and where I can change potential mappings. Also tried Dr. Google but had zero luck. Anybody any suggestion what to try - at least try. I am currently 100% lost.

**Comment by Guvalle** [comment]
> 
> 
> @jschnorr Looking closer at your process list, it sounds like you're telling debugpy to listen on 127.0.0.1. This would prevent connections from other machines.
> 
> Most likely, you're only specifying the port and not the host in `--listen` or `debugpy.listen()`, whichever one you're using - it defaults to 127.0.0.1 for host, for security reasons. To make it listen on all interfaces, do:
> 
> ```python
> debugpy.listen(('0.0.0.0', 5678))
> ```
> 
> or via CLI
> 
> ```shell
> debugpy --listen 0.0.0.0:5678
> ```
> 
> But note that this should only be done if your network is secure, because anybody who is able to connect to that port, can effectively run arbitrary code on that machine!

@int19h Thank you so much. I've been looking for a solution to this problem for hours and found nothing useful until I read your suggestion. Thank you very much, I changed my "debugpy.listen(5678)" to "debugpy.listen(('0.0.0.0', 5678))" and it worked

**Comment by agribot2** [comment]
Hi 
Can you please provide more details on how to use breakpoints in this situation ( remote debugging python ) 
 [https://docs.microsoft.com/de-de/visualstudio/python/debugging-python-code-on-remote-linux-machines?view=vs-2019](url)

I run the code on my remote machine ( a raspberry pi 4 board)

In Visual Studio 2019 ( on my windows 10 laptop ) it displays my remote entries in the VS Output window, but it does not display any variables in the Locals window.
Can you please provide some examples of how to use breakpoints, watch variables, status, etc.

Many thanks

**Comment by int19h** [comment]
@agribot2 It really should "just work" when you pause your app, either by using a breakpoint, or manually via the corresponding command on the Debug menu or toolbar. If not, can you please file a separate issue?
